linux-5.10.35.patch

```diff
@@ -2040,6 +2030,63 @@ index 45fb450b4..48da5be43 100644
  	}
  
  	offset++;
+diff --git a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
+index 97975bed4..b2feccb2c 100644
+--- a/net/ipv4/ip_output.c
++++ b/net/ipv4/ip_output.c
+@@ -701,8 +701,40 @@ struct sk_buff *ip_frag_next(struct sk_buff *skb, struct ip_frag_state *state)
+ 		len &= ~7;
+ 	}
+ 
++#ifdef CONFIG_SECURITY_TEMPESTA
++/*
++ * Since Tempesta FW tries to reuse incoming SKBs containing the response
++ * from the backend, sometimes we might encounter an SKB with quite a small
++ * head room, which is not big enough to accommodate all the transport headers
++ * and TLS overhead.
++ * It usually the case when working over loopback, tun/tap, bridge or similar
++ * interfaces with small MTU. The issue is specific to aforementioned ifaces
++ * because the outgoing SKB would be injected back to the stack.
++ * In order not to reallocate sk_buffs' headroom on on RX path,
++ * allocate and reserve a little bit more memory on TX path.
++ * Even though it would introduce some memory overhead, it's still
++ * cheaper than doing transformation.
++ *
++ * It seems like no such actions are required for IPv6 counterparts:
++ * ip6_fragment() / ip6_frag_next() due to the fact that the
++ * lowest acceptable MTU (1280) is sufficient to fit all the headers.
++ *
++ * When receiving SKBs from the outter world, the NIC driver should
++ * allocate and reserve all necessary space by itself.
++ */
++
++/* Redefine this constant from include/net/tcp.h here, not to bring
++ * the whole header */
++#define MAX_TCP_HEADER	L1_CACHE_ALIGN(128 + MAX_HEADER)
++#endif
++
+ 	/* Allocate buffer */
++#ifdef CONFIG_SECURITY_TEMPESTA
++	skb2 = alloc_skb(len + state->hlen + state->ll_rs
```

Comment:
We can allocate only MAX_TCP_HEADER here and reserve MAX_TCP_HEADER - len - state->hlen

Reply:
I believe we can't do this. For instance, if `len + state->hlen` above than `MAX_TCP_HEADER` we will reserve for headroom only `state->ll_rs` that can be not enough to contain all necessary TCP + TLS headers. At this place we can't predict required size of headers. If we could determine whether this skb will be reused, then we could avoid reserving MAX_TCP_HEADER and use only `ll` room.